Topic-4-Javascript-Desing-Patterns/2/2.js

```diff
@@ -0,0 +1,53 @@
+class EventEmiter {
+    constructor () {
+        this.events = {};
+    }
+    on (eventName , callback) {
+        if ( this.events[eventName]) {
+            this.events[eventName].push(callback);
+        } else {
+            this.events[eventName] = [callback];
+        }        
+    }
+    emit (eventName) {
+        this.events[eventName].forEach(callback =>
+            callback());       
+    }
+    off (eventName , callback) {
+        let newEvents = this.events[eventName].filter(c => c !== callback);
+        this.events[eventName] = newEvents;
+    }
+    
+}
+
+class Movie extends EventEmiter{
+
+    constructor (title,year,duration) {
+        super();
+        this.title = title;
+        this.year = year;
+        this.duration = duration;
+    }
+    play() {
```

Comment:
the _on_ method is defined as `on (eventName , callback)` here you are calling `this.on(undefined)` because you're invoking the console.log instead of providing it as a callback. In another hand, every time you execute your movie method play you will run a subscription. Instead, you need to:

1. Use the event emitter methods to subscribe a callback to the play, pause and resume events
2. In every method, you need to **publish** (_emit_) the corresponding event.